src/coprocessor/dag/executor/scanner.rs

```diff
@@ -116,34 +134,41 @@ impl<S: Snapshot> Scanner<S> {
             );
         }
 
-        self.seek_key = match (self.scan_mode, self.scan_on) {
-            (ScanMode::Forward, _) => util::prefix_next(&key), // Where is this thing useful?????
-            (ScanMode::Backward, ScanOn::Table) => box_try!(truncate_as_row_key(&key)).to_vec(),
-            (ScanMode::Backward, ScanOn::Index) => key.clone(),
-            _ => unreachable!(),
-        };
+        {
+            let seek_key_slice = match (self.scan_mode, self.scan_on) {
+                (ScanMode::Forward, _) | (ScanMode::Backward, ScanOn::Index) => key.as_slice(),
+                (ScanMode::Backward, ScanOn::Table) => box_try!(truncate_as_row_key(&key)),
+                _ => unreachable!(),
+            };
+            self.seek_key.clear();
```

Comment:
set_len is more fast.

Reply:
Is it ok to write unsafe code here?

Reply:
Bellowing is the code, you can see the difference obviously.

    pub unsafe fn set_len(&mut self, len: usize) {
        self.len = len;
    }

pub fn clear(&mut self) {
        self.truncate(0)
    }
pub fn truncate(&mut self, len: usize) {
        unsafe {
            // drop any extra elements
            while len < self.len {
                // decrement len before the drop_in_place(), so a panic on Drop
                // doesn't re-drop the just-failed value.
                self.len -= 1;
                let len = self.len;
                ptr::drop_in_place(self.get_unchecked_mut(len));
            }
        }
    }